app/gui2/src/util/ffi.ts

```diff
@@ -0,0 +1,406 @@
+import init, { parse_to_json } from '../../rust-ffi/pkg/rust_ffi'
+import type { NonEmptyArray, Opt } from './opt'
+
+const _wasm = await init()
+
+export function parseEnso(code: string): ast.Tree {
+  const json = parse_to_json(code)
+  return JSON.parse(json)
+}
+
+export module ast {
+  export interface Tree {
+    span: Span
+    variant: Variant
+  }
+
+  export type Token =
+    | Token.AutoScope
+    | Token.CloseSymbol
+    | Token.Digits
+    | Token.Ident
+    | Token.Newline
+    | Token.NumberBase
+    | Token.OpenSymbol
+    | Token.Operator
+    | Token.TextEnd
+    | Token.TextEscape
+    | Token.TextSection
+    | Token.TextStart
+    | Token.Wildcard
+
+  module Token {
+    interface TokenBase<T> {
+      left_offset: Offset
+      code: Code
+      variant: T
+    }
+
+    export type AutoScope = TokenBase<{}>
+    export type CloseSymbol = TokenBase<{}>
```

Comment:
typescript is structurally typed, so this means `AutoScope` and `CloseSymbol` are equivalent types - it is impossible to tell them apart. you may or may not want to add brands

Reply:
Right. I was kinda surprised that those are not already distinguishable after deserialization. They really are the same type now. Ideally we would just fix rust serde definition to make them actually distinct. But I don't want to mess with the parser too much now.